File: companies/Nutanix/189.rotate-array.cpp
#include <algorithm>
#include <array>
#include <cctype>
#include <climits>
#include <cstdlib>
#include <functional>
#include <numeric>
#include <queue>
#include <sstream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>
using namespace std;


class Solution {
public:
    void rotate(vector<int>& nums, int k) {
        /*
        Approach: reverse the whole array, reverse the first k elements, then
        reverse the remaining n-k elements. This moves the last k elements to the
        front while preserving both groups' internal order.

        Complexity: O(n) time, O(1) space.
        */
        int n = nums.size();
        k %= n;
        reverse(nums.begin(), nums.end());
        reverse(nums.begin(), nums.begin() + k);
        reverse(nums.begin() + k, nums.end());
    }
};

File: Texas Instruments/189.rotate-array.go
package leetcode

// Rotate189 rotates right by k using three reversals: reverse whole array,
// reverse the first k values, then reverse the remainder.
//
// Time: O(n). Space: O(1).
func Rotate189(nums []int, k int) {
	n := len(nums)
	if n == 0 {
		return
	}
	k %= n
	rev := func(l, r int) {
		for l < r {
			nums[l], nums[r] = nums[r], nums[l]
			l++
			r--
		}
	}
	rev(0, n-1)
	rev(0, k-1)
	rev(k, n-1)
}
